### PR TITLE
Add an architecture overview to the developer guide

### PR DESCRIPTION
**Pull request by mxm**
This is the beginning of a rough overview of CrateDB's architecture. It is only
a starting point and I hope we can expand it over time. The goal is that people
new to the project can get an idea of what technology we use and how the project
is structured.

Feedback welcome.

### CODE REVIEW

File: devs/docs/architecture.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+- `Planner`
+- `LogicalPlan` / `Plan`
+- `NodeOperationTree`
+- `ExecutionPhasesTask`
+- `BatchIterator`
+
+Input
+.....
+
+When you first use CrateDB, you probably want to check out its Web
```
Comment (mfussenegger) [inline devs/docs/architecture.rst]: I'd assume that the reader has some familiarity with CrateDB from a user perspective. Otherwise the gap here from basic introduction to code seem quite large

File: devs/docs/architecture.rst [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+Class entry points:
+
+- `PostgresWireProtocol`
+- `CrateRestMainAction`
+
+Transport
+..........
+
+Transport denotes the communication between nodes in a CrateDB cluster. CrateDB
```
Comment (mfussenegger) [inline devs/docs/architecture.rst]: This part seems quite minimal - maybe ask someone more unfamiliar if these few pointers are already helpful?

File: devs/docs/architecture.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+Class entry points:
+
+- `TransportRequest`
+- `TransportJobAction`
+
+Storage
+........
+
+CrateDB enables you to store your data in tables like you would in a traditional
+SQL database with a strict schema. Additionally, you can dynamically adjust the
```
Comment (mfussenegger) [inline devs/docs/architecture.rst]: `adjust` might imply removing columns, which we currently don't support. I think `extend` would be more accurate

File: devs/docs/architecture.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+schema or store JSON objects inside columns which you can also query. Data is
+clustered by a column or expression to distribute the data. Partitioned tables
+allow you to further distribute and split up your data using other columns or
+expressions.
+
+To be able to retrieve data efficiently and perform aggregations, CrateDB uses
+Lucene for indexing and storing data. Lucene itself stores a document with the
```
Comment (mfussenegger) [inline devs/docs/architecture.rst]: I think this is a bit misleading. There are several different kind of storage mechanism which are all in use for different purposes. Maybe we should just link to the Lucene documentation about the storage formats?

File: devs/docs/architecture.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Class entry points:
+
+- `LuceneQueryBuilder`
+- `LuceneBatchIterator`
+
+Enterprise
+..........
+
+The enterprise version of CrateDB contains features which are additions to the
```
Comment (mfussenegger) [inline devs/docs/architecture.rst]: Is this relevant from a developer / architecture point of view?

File: devs/docs/architecture.rst [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+............
+
+CrateDB's heart consists of a SQL Engine which takes care of parsing SQL
+statements and executing them in the cluster.
+
+The engine is comprised of the following components:
+
+1. Parser: Breaks down the SQL statement into its components.
+2. Analyzer: Performs semantic processing of the statement including
+   verification, type annotation, and optimization.
```
[the remark this answers is not in the document]
Comment (mxm) [reply devs/docs/architecture.rst]: 👍

File: devs/docs/architecture.rst [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+3. Planner: Builds an execution plan from the analyzed statement.
+4. Executor: Executes the plan in the cluster and collects results.
+
+Class entry point:
+
+- `SqlParser`
+- `RelationAnalyzer`
```
[the remark this answers is not in the document]
Comment (mxm) [reply devs/docs/architecture.rst]: Will replace with Analyzer.

>I'm also wondering how we ensure to keep this up to date?

We just update it once it it outdated. Also, people might read it and open PRs to fix it. Note that eventual inconsistencies here don't mean that the document is invalid. It still provides hints for newcomers.

File: devs/docs/architecture.rst (hunk 1 above)
Comment (mxm) [reply devs/docs/architecture.rst]: Yes, it is very hard to imagine someone contributing without having used CrateDB at all.

File: devs/docs/architecture.rst (hunk 2 above)
Comment (mxm) [reply devs/docs/architecture.rst]: Yes, need to expand on this further. Ideally, someone else could add a paragraph here.

File: devs/docs/architecture.rst (hunk 3 above)
Comment (mxm) [reply devs/docs/architecture.rst]: 👍

File: devs/docs/architecture.rst (hunk 4 above)
Comment (mxm) [reply devs/docs/architecture.rst]: On a high-level Lucene always stores documents. After this, I mention there are various storage options which help us to retrieve or aggregate data efficiently.

https://lucene.apache.org/core/3_0_3/fileformats.html#Index%20File%20Formats

File: devs/docs/architecture.rst (hunk 5 above)
Comment (mxm) [reply devs/docs/architecture.rst]: My point here was to state that CrateDB is a functional DB without the enterprise parts. Any ideas what to add here or should we just remove the paragraph?

File: devs/docs/architecture.rst (hunk 1 above)
Comment (mfussenegger) [reply devs/docs/architecture.rst]: That's why I'd even shorten the web-ui introduction here. Instead I'd state in the introduction that it's expected that the reader is familiar with the functionality of CrateDB and link to the user documentation.

Anyway - these are just some thoughts / inputs, because I felt like being picked up very gentle just to be dropped into the deep water afterwards. The mental gap being "There is an admin ui" and "here are two components which are part of the protocol implementation" is rather large. But to be honest I've also no good idea how to soften this gap without getting lost in all the details.

This applies generally to the whole document as it is. Personally I feel that the narrative contains too many big leaps. But I'm probably also not the best person to make this judgement.

I think it's fine to get this in if we see this is a starting point. But if you want to iterate on it I'd recommend to have some more people read it and talk to them to see what kind of questions pop up.

File: devs/docs/architecture.rst (hunk 4 above)
Comment (mfussenegger) [reply devs/docs/architecture.rst]: I guess this was just me feeling that this is an over-simplification that could lead people to form incorrect conclusions. E.g. that indexing speeds up the retrieval - the indices are used to query/determine the right documents, but for the value retrieval it's actually not used (at least not in the sense that it might be used in a traditional DBMS like postgres)

But I can see why you want to keep it at this level. (Though I'd still add a link to Lucene, though ideally to a more recent version)

File: devs/docs/architecture.rst (hunk 1 above)
Comment (mxm) [reply devs/docs/architecture.rst]: I think it is fine to have leaps in this document. It is just to have an overview of what components exists, alongside with some pointers to the implementation.

I've tried to get feedback from more people. Will try harder ;)

Comment (mxm) [reply devs/docs/architecture.rst]: I've removed the admin ui information and added a pointer to the getting started section in the beginning.

**Comment by mxm** [comment]
I've added some more people to this PR. Would be nice to collect some more feedback or ideas from you. Do you find this helpful? What are you missing. Don't hesitate to comment on anything.

File: devs/docs/architecture.rst [hunk 8]
```diff
@@ -0,0 +1,98 @@
+=====================
+CrateDB Architecture
+=====================
+
+This documents assumes that you are familiar with using CrateDB. If you haven't
```
Comment (emilywoods) [inline devs/docs/architecture.rst]: Can remove `s` in `documents`

File: devs/docs/architecture.rst [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+CrateDB enables you to store your data in tables like you would in a traditional
+SQL database with a strict schema. Additionally, you can dynamically extend the
+schema or store JSON objects inside columns which you can also query. Data is
+clustered by a column or expression to distribute the data. Partitioned tables
+allow you to further distribute and split up your data using other columns or
+expressions.
+
+To be able to retrieve data efficiently and perform aggregations, CrateDB uses
+Lucene for indexing and storing data. Lucene itself stores a document with the
+contents of each row. Retrieval is really efficient because the fields of the
```
Comment (emilywoods) [inline devs/docs/architecture.rst]: This might be a bit pedantic, but I think the word `really` could be left out here. I don't think words like 'very' or 'really' add much value unless you have something concrete to compare it against.

File: devs/docs/architecture.rst (hunk 8 above)
Comment (mxm) [reply devs/docs/architecture.rst]: Thanks for spotting!

File: devs/docs/architecture.rst (hunk 9 above)
Comment (mxm) [reply devs/docs/architecture.rst]: Good point.

File: devs/docs/architecture.rst (hunk 5 above)
Comment (Autophagy) [reply devs/docs/architecture.rst]: I think removing it would be best. It's not really relevant from an architectural point of view, imo

File: devs/docs/architecture.rst [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+Class entry points:
+
+- `CrateRestMainAction`
+- `PostgresWireProtocol`
+
+Transport
+..........
+
+Transport denotes the communication between nodes in a CrateDB cluster. CrateDB
+uses Netty and Elasticsearch to transfer data between nodes.
```
Comment (Autophagy) [inline devs/docs/architecture.rst]: Might it be worth briefly mentioning what *kinds* of communication occur on the transport layer? Like synchronising the cluster start, redistributing shards, node discovery, etc?

File: devs/docs/architecture.rst (hunk 5 above)
Comment (mxm) [reply devs/docs/architecture.rst]: Good point. Removed the paragraph.

File: devs/docs/architecture.rst (hunk 10 above)
Comment (mxm) [reply devs/docs/architecture.rst]: 👍 I've pushed a fixup.

File: devs/docs/architecture.rst [hunk 11]
```diff
@@ -0,0 +1,95 @@
+=====================
+CrateDB Architecture
+=====================
+
+This document assumes that you are familiar with using CrateDB. If you haven't
+used CrateDB before, please check out the [Getting
```
Comment (Autophagy) [inline devs/docs/architecture.rst]: The getting started link here is in markdown format, not in restructured text :wink:

Comment (mxm) [reply devs/docs/architecture.rst]: 👍  fixed

File: devs/docs/architecture.rst [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+- `LogicalPlan` / `Plan`
+- `NodeOperationTree`
+- `ExecutionPhasesTask`
+- `BatchIterator`
+
+Input
+.....
+
+CrateDB accepts input from an HTTP/REST interface as well as from a Postgres
+compatible protocol. The Admin UI, crash and some of the connectors use the HTTP
```
Comment (turbo-ele) [inline devs/docs/architecture.rst]: what about mqtt?

File: devs/docs/architecture.rst [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+- `TransportRequest`
+- `TransportJobAction`
+
+Storage
+........
+
+CrateDB enables you to store your data in tables like you would in a traditional
+SQL database with a strict schema. Additionally, you can dynamically extend the
+schema or store JSON objects inside columns which you can also query. Data is
+clustered by a column or expression to distribute the data. Partitioned tables
```
Comment (matriv) [inline devs/docs/architecture.rst]: Maybe you can mention the hierarchy:
```
Table
Pt1                                               Pt2 ....                      PtN
Pt1_sh1  Pt1_sh2...  Pt1_shM              
```

File: devs/docs/architecture.rst (hunk 12 above)
Comment (mxm) [reply devs/docs/architecture.rst]: Good point. I now mention it below that paragraph.

File: devs/docs/architecture.rst (hunk 13 above)
Comment (mxm) [reply devs/docs/architecture.rst]: That's a great idea. I would like to expand on how tables are stored. Have to think about how I do this without introducing too much detail here.

Comment (mxm) [reply devs/docs/architecture.rst]: I've actually added a table to visualize the table/shards/partitions.

File: devs/docs/architecture.rst [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+allow you to further distribute and split up your data using other columns or
+expressions.
+
+A table corresponds to an Elasticsearch index. An index may have one or multiple
+shards which represent the slices of a table which are stored across the
+cluster. If a table is partitioned, each partition has its own index.
```
Comment (matriv) [inline devs/docs/architecture.rst]: Would rephrase it to say that each partition is (or corresponds) an Elasticsearch index. The word `has` sounds weird to me :-)

File: devs/docs/architecture.rst [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+schema or store JSON objects inside columns which you can also query. Data is
+clustered by a column or expression to distribute the data. Partitioned tables
+allow you to further distribute and split up your data using other columns or
+expressions.
+
+A table corresponds to an Elasticsearch index. An index may have one or multiple
+shards which represent the slices of a table which are stored across the
+cluster. If a table is partitioned, each partition is a separate index.
+
+The following table illustrates the table shows the relationship between table,
```
Comment (matriv) [inline devs/docs/architecture.rst]: please check again the sentence, use either illustrates or shows and you have the table twice  :-)

**Comment by mxm** [comment]
Thanks :)
